expression/collation.go

```diff
@@ -322,7 +317,11 @@ func safeConvert(ctx sessionctx.Context, ec *ExprCollation, args ...Expression)
 			if err != nil {
 				return false
 			}
-			if !isNull && !isValidString(str, ec.Charset) {
+			// if value is NULL or binary string, just skip it.
+			if isNull || types.IsBinaryStr(c.GetType()) {
+				continue
+			}
```

Comment:
We can move `types.IsBinaryStr(c.GetType())` to the beginning of this loop to avoid unnecessary `EvalString`.

Reply:
@xiongjiwei Please address this comment.